plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,249 @@
+		s.Queries[i].tagFilter, err = compileTagFilter(q)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func compileTagFilter(q Query) (filter.Filter, error) {
+	tagFilter, err := filter.NewIncludeExcludeFilterDefaults(q.TagPropertiesInclude, nil, false, false)
+	if err != nil {
+		return nil, fmt.Errorf("creating tag filter failed: %v", err)
+	}
+	return tagFilter, nil
+}
+
+func buildWqlStatements(s *Wmi) {
+	for i, q := range s.Queries {
+		s.Queries[i].query = buildWqlStatement(q)
+	}
+}
+
+// build a WMI query from input configuration
+func buildWqlStatement(q Query) string {
+	wql := fmt.Sprintf("SELECT %s FROM %s", strings.Join(q.Properties, ", "), q.ClassName)
+	if len(q.Filter) > 0 {
+		wql = fmt.Sprintf("%s WHERE %s", wql, q.Filter)
+	}
+	return wql
+}
+
+func (q *Query) doQuery(acc telegraf.Accumulator) error {
+	// The only way to run WMI queries in parallel while being thread-safe is to
+	// ensure the CoInitialize[Ex]() call is bound to its current OS thread.
+	// Otherwise, attempting to initialize and run parallel queries across
+	// goroutines will result in protected memory errors.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	tags := map[string]string{}
+	fields := map[string]interface{}{}
+
+	// init COM
+	if err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED); err != nil {
+		oleCode := err.(*ole.OleError).Code()
+		if oleCode != ole.S_OK && oleCode != sFalse {
+			return err
+		}
+	}
+	defer ole.CoUninitialize()
+
+	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
+	if err != nil {
+		return err
+	} else if unknown == nil {
+		return fmt.Errorf("failed to create WbemScripting.SWbemLocator, is WMI broken?")
+	}
+	defer unknown.Release()
+
+	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
+	if err != nil {
+		return fmt.Errorf("failed to QueryInterface: %v", err)
+	}
+	defer wmi.Release()
+
+	// service is a SWbemServices
+	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer", nil, q.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed calling method ConnectServer: %v", err)
+	}
+	service := serviceRaw.ToIDispatch()
+	defer func() { _ = serviceRaw.Clear() }()
+
+	// result is a SWBemObjectSet
+	resultRaw, err := oleutil.CallMethod(service, "ExecQuery", q.query)
+	if err != nil {
+		return fmt.Errorf("failed calling method ExecQuery for query %s: %v", q.query, err)
+	}
+	result := resultRaw.ToIDispatch()
+	defer func() { _ = resultRaw.Clear() }()
+
+	count, err := oleInt64(result, "Count")
+	if err != nil {
+		return fmt.Errorf("failed getting Count: %v", err)
+	}
+
+	for i := int64(0); i < count; i++ {
+		// item is a SWbemObject
+		itemRaw, err := oleutil.CallMethod(result, "ItemIndex", i)
+		if err != nil {
+			return fmt.Errorf("failed calling method ItemIndex: %v", err)
+		}
+
+		outerErr := func() error {
+			item := itemRaw.ToIDispatch()
+			defer item.Release()
+
+			for _, wmiProperty := range q.Properties {
+				innerErr := func() error {
```

Comment:
Same here. Please make this an explicitly declared function.

Reply:
Same as previous occurrence. Removed func() declaration and reduced nesting.